reframe/frontend/statistics.py

```diff
@@ -1,89 +1,103 @@
-
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+    def __init__(self):
+        self._tasks = [[]]
+        self._current_run = 0
 
     def __repr__(self):
         return debug.repr(self)
 
-    def num_failures(self, partition=None):
-        num_fails = 0
-        if partition:
-            num_fails += len([
-                t for t in self._tasks_bypart[partition] if t.failed
-            ])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_fails += len([t for t in tclist if t.failed])
-
-        return num_fails
-
-    def num_failures_stage(self, stage):
-        num_fails = 0
-        for tclist in self._tasks_bypart.values():
-            num_fails += len([t for t in tclist if t.failed_stage == stage])
-
-        return num_fails
-
-    def num_cases(self, partition=None):
-        num_cases = 0
-        if partition:
-            num_cases += len(self._tasks_bypart[partition])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_cases += len(tclist)
-
-        return num_cases
+    def next_run(self, current_run):
+        if current_run != self._current_run + 1:
+            raise ReframeFatalError('current_run variable out of sync')
+
+        self._current_run = current_run
+        self._tasks.append([])
+
+    def add_task(self, task):
+        self._tasks[self._current_run].append(task)
+
+    def _get_tasks(self, run):
+        try:
+            return self._tasks[run]
+        except IndexError:
+            raise StatisticsError('no such run: %s' % run)
+
+    def num_failures(self, run=-1):
+        return len([t for t in self._get_tasks(run) if t.failed])
+
+    def num_cases(self, run=-1):
+        return len(self._get_tasks(run))
+
+    def tasks_failed(self, run=-1):
+        return [t for t in self._get_tasks(run) if t.failed]
+
+    def retry_report(self):
+        # Return an empty report if no retries were done.
+        if not self._current_run:
+            return ''
+
+        line_width = 78
+        report = [line_width * '=']
+        report.append('SUMMARY OF RETRIES')
+        report.append(line_width * '-')
+        messages = {}
+        for run in range(1, len(self._tasks)):
+            for t in self._get_tasks(run):
+                # Overwrite entry from previous run if available
+                messages[t.check.info()] = (
+                    '  * Test %s was retried %s time(s) and %s.' %
+                    (t.check.info(), run, 'failed' if t.failed else 'passed')
+                )
+
+        for key in sorted(messages.keys()):
```

Comment:
What's the purpose of sorting here?

Reply:
To have the output sorted by the test name (and the other elements that appear in the key) as
```
Test A ...
Test A ...
Test A ...
Test C ...
Test D ...
Test D ...
```
 Else we could have an output like:
```
Test D ...
Test A ...
Test C ...
Test A ...
Test D ...
Test A ...
```

Reply:
Ok, leave it sorted then.

Reply:
Done.